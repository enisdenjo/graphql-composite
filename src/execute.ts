import { ExecutionResult, GraphQLError } from 'graphql';
import { jsonToGraphQLQuery } from 'json-to-graphql-query';
import getAtPath from 'lodash.get';
import setAtPath from 'lodash.set';
import { GatherPlan, GatherPlanResolver } from './gather.js';

export interface ExecutionContext {
  /** Fetch getter by {@link SchemaPlanSource.id} used for {@link SchemaPlanFetchResolver}. */
  getFetch: (sourceId: string) => typeof fetch;
}

export async function execute(
  ctx: ExecutionContext,
  plan: GatherPlan,
  variables: Record<string, unknown>,
): Promise<ExecutionResult & { explain: ResolverExplain[] }> {
  const result: ExecutionResult = {};
  // TODO: batch resolvers going to the same source
  const explain = await Promise.all(
    plan.operations.flatMap((o) =>
      o.resolvers.map((r) =>
        executeResolver(ctx, r, variables, [r.typeOrField], result),
      ),
    ),
  );
  return { ...result, explain };
}

export interface ResolverExplain {
  path: (string | number)[];
  query: string;
  variables: Record<string, unknown>;
  data?: unknown;
  errors?: ReadonlyArray<GraphQLError>;
  exports: GatherPlanResolver['exports'];
  includes: ResolverExplain[];
}

async function executeResolver(
  ctx: ExecutionContext,
  resolver: GatherPlanResolver,
  /** The variables for the operation. Those that the user provides. */
  operationVariables: Record<string, unknown> | null,
  /**
   * The path in the {@link resultRef} this gather is resolving.
   */
  path: (string | number)[],
  /**
   * Mutated argument that references the final result,
   * ready to be passed back to the user.
   */
  resultRef: ExecutionResult,
): Promise<ResolverExplain> {
  const { getFetch } = ctx;

  const query = buildResolverQuery(resolver);
  const variables =
    operationVariables ||
    Object.values(resolver.variables).reduce(
      (agg, { name, select }) => ({
        ...agg,
        [name]: getAtPath(resultRef.data, [...path, select]),
      }),
      {},
    );
  const res = await getFetch(resolver.source)(resolver.url, {
    method: 'POST',
    headers: {
      'content-type': 'application/json',
      accept: 'application/graphql-response+json, application/json',
    },
    body: JSON.stringify({
      query,
      // TODO: actual operation name
      operationName: null,
      variables,
    }),
  });
  if (!res.ok) {
    const err = new Error(
      `${res.status} ${res.statusText}\n${await res.text()}`,
    );
    err.name = 'ResponseError';
    throw err;
  }

  const result: ExecutionResult = await res.json();
  if (result.errors?.length) {
    // TODO: should throw instead?
    // stop immediately on errors, no need to traverse futher
    resultRef.errors = result.errors;
    return {
      path,
      query,
      variables,
      ...result,
      exports: [], // TODO: nothing's exported because of the fail, should we popoulate regardless?
      includes: [],
    };
  }
  if (!result.data) {
    const err = new Error(
      'GraphQL execution result has no errors but does not contain data',
    );
    err.name = 'GraphQLExecutionNoData';
    throw err;
  }

  if (!resultRef.data) {
    resultRef.data = {};
  }
  for (const exportPath of resolver.exports.map((e) => e.split('.'))) {
    const lastKey = exportPath[exportPath.length - 1];
    const valBeforeLast =
      exportPath.length > 1
        ? getAtPath(
            result.data[
              // TODO: we only assume the root field is where the export fragment is, but isnt always true
              Object.keys(result.data)[0]!
            ],
            exportPath.slice(0, exportPath.length - 1),
          )
        : null;

    if (Array.isArray(valBeforeLast)) {
      // set key in each item of the array
      for (let i = 0; i < valBeforeLast.length; i++) {
        setAtPath(
          resultRef.data,
          [
            ...path,
            ...exportPath.slice(0, exportPath.length - 1)!,
            i,
            lastKey!,
          ],
          getAtPath(valBeforeLast[i], [lastKey!]),
        );
      }
    } else {
      setAtPath(
        resultRef.data,
        [...path, ...exportPath],
        getAtPath(
          result.data[
            // TODO: we only assume the root field is where the export fragment is, but isnt always true
            Object.keys(result.data)[0]!
          ],
          exportPath,
        ),
      );
    }
  }

  return {
    path,
    query,
    variables,
    ...result,
    exports: resolver.exports,
    // TODO: batch resolvers going to the same source
    includes: await Promise.all(
      Object.entries(resolver.includes).flatMap(([field, resolver]) => {
        const val = getAtPath(result.data, [...path, field]);
        if (Array.isArray(val)) {
          // if the include points to an array, we need to gather resolve each item
          return val.map((_, i) =>
            executeResolver(
              ctx,
              resolver,
              null,
              [...path, field, i],
              resultRef,
            ),
          );
        }
        return executeResolver(
          ctx,
          resolver,
          null,
          [...path, field],
          resultRef,
        );
      }),
    ),
  };
}

export function buildResolverQuery(
  resolver: Pick<GatherPlanResolver, 'operation' | 'type' | 'exports'>,
) {
  let query = resolver.operation;
  query += ` fragment export on ${resolver.type} { `;
  const obj = {};
  for (const path of resolver.exports) {
    setAtPath(obj, path, true);
  }
  query += jsonToGraphQLQuery(obj);
  query += ' }';
  return query;
}
