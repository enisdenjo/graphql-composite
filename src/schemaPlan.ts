import {
  ConstDirectiveNode,
  GraphQLField,
  GraphQLObjectType,
  GraphQLSchema,
  Kind,
  OperationTypeNode,
} from 'graphql';

export interface SchemaPlan {
  operations: Record<
    string /* OperationDefinitionNode.name */,
    SchemaPlanOperation
  >;
  compositeTypes: Record<
    string /* SchemaPlanCompositeType.name */,
    SchemaPlanCompositeType
  >;
}

export interface SchemaPlanOperation {
  name: OperationTypeNode;
  fields: Record<
    string /* SchemaPlanOperationField.name */,
    SchemaPlanOperationField
  >;
}

export interface SchemaPlanOperationField {
  name: string;
  resolvers: Record<
    string /* SchemaPlanSource.id */,
    SchemaPlanSource & SchemaPlanResolver // TODO: operation field must always have a resolver?
  >;
}

export interface SchemaPlanCompositeType {
  name: string;
  resolvers: Record<
    string /* SchemaPlanSource.id */,
    SchemaPlanSource & SchemaPlanResolver // TODO: type can only have one resolver?
  >;
  fields: Record<
    string /* SchemaPlanCompositeTypeField.name */,
    SchemaPlanCompositeTypeField
  >;
}

export interface SchemaPlanCompositeTypeField {
  name: string;
  sources: Record<
    string /* SchemaPlanSource.id */,
    (SchemaPlanSource & SchemaPlanResolver) | SchemaPlanSource // a type field may not have a resolver, assuming it's in available in the type
  >;
}

export interface SchemaPlanSource {
  /** Unique ID of the source. Usually the subgraph name. */
  id: string;
  /** Name of the type or field in the source subgraph. */
  name: string;
}

export type SchemaPlanResolver = SchemaPlanFetchResolver; // TODO: other kinds

export interface SchemaPlanFetchResolver {
  kind: 'fetch';
  operation: string;
  variables: Record<
    string /* SchemaPlanResolverVariable.name */,
    SchemaPlanResolverVariable
  >;
  /** Location of the source to be used by the resolver. */
  url: string;
}

interface SchemaPlanResolverVariable {
  name: string;
  /** Which field in the type to use as this variable. */
  select: string;
}

export function planSchema(schema: GraphQLSchema): SchemaPlan {
  const plan: SchemaPlan = { operations: {}, compositeTypes: {} };

  const types = schema.getTypeMap();

  for (const [name, type] of Object.entries(types)) {
    if (name.startsWith('__')) {
      // ignore introspection types
      continue;
    }
    if (!(type instanceof GraphQLObjectType)) {
      // ignore non object types
      continue;
    }

    if (
      type.name === 'Query' ||
      type.name === 'Mutation' ||
      type.name === 'Subscription'
    ) {
      // operations
      const name = type.name.toLowerCase() as OperationTypeNode;
      const operation: SchemaPlanOperation = {
        name: name,
        fields: {},
      };
      for (const [name, field] of Object.entries(type.getFields())) {
        const fieldPlan: SchemaPlanOperationField = {
          name,
          resolvers: {},
        };
        for (const source of getSourcesFromDirectives(type, field)) {
          const resolver = getResolverForSourceFromDirectives(
            type,
            field,
            source,
          );
          if (!resolver) {
            throw new Error(
              `Operation field "${field.name}" must have a resolver`,
            );
          }
          fieldPlan.resolvers[source.id] = {
            ...source,
            ...resolver,
          };
        }
        operation.fields[fieldPlan.name] = fieldPlan;
      }
      plan.operations[name] = operation;
      continue;
    }

    // compositeTypes
    const compositeTypePlan: SchemaPlanCompositeType = {
      name: type.name,
      fields: {},
      resolvers: {},
    };
    for (const source of getSourcesFromDirectives(type, null)) {
      const resolver = getResolverForSourceFromDirectives(type, null, source);
      compositeTypePlan.resolvers[source.id] = { ...source, ...resolver };
    }
    for (const [name, field] of Object.entries(type.getFields())) {
      const fieldPlan: SchemaPlanCompositeTypeField = {
        name,
        sources: {},
      };
      for (const source of getSourcesFromDirectives(type, field)) {
        const resolver = getResolverForSourceFromDirectives(
          type,
          field,
          source,
        );
        fieldPlan.sources[source.id] = { ...source, ...resolver };
      }
      compositeTypePlan.fields[fieldPlan.name] = fieldPlan;
    }
    plan.compositeTypes[type.name] = compositeTypePlan;
  }

  return plan;
}

/**
 * Gets the sources of either the `field`, if provided, or the `type`.
 * Both arguments are used to better the error messages.
 */
function getSourcesFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
): SchemaPlanSource[] {
  const fieldOrType = field || type;
  const sources: SchemaPlanSource[] = [];
  for (const sourceDirective of fieldOrType.astNode?.directives?.filter(
    (d) => d.name.value === 'source',
  ) || []) {
    // TODO: make sure source doesnt already exist
    sources.push({
      id: mustGetStringArgumentValue(type, field, sourceDirective, 'subgraph'),
      name: mustGetStringArgumentValue(type, field, sourceDirective, 'name'),
    });
  }
  return sources;
}

/**
 * Gets the resolver of the source in either the `field`, if provided, or the `type`.
 * Both arguments are used to better the error messages.
 */
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: null,
  source: SchemaPlanSource,
): SchemaPlanResolver;
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown>,
  source: SchemaPlanSource,
): SchemaPlanResolver | null;
function getResolverForSourceFromDirectives(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
  source: SchemaPlanSource,
): SchemaPlanResolver | null {
  const fieldOrType = field || type;
  const directives = fieldOrType.astNode?.directives || [];
  const resolverDirs = directives.filter(
    (d) =>
      d.name.value === 'resolver' &&
      mustGetStringArgumentValue(type, field, d, 'subgraph') === source.id,
  );
  const resolverDir = resolverDirs[0];
  if (!resolverDir) {
    if (field === fieldOrType) {
      // sources on fields must not always have a resolver
      return null;
    }
    throw new Error(
      `source "${source.id}" on type "${type.name}" has no resolver directive`,
    );
  }
  if (resolverDirs.length > 1) {
    throw new Error(
      `source "${source.id}" on type "${type.name}" has multiple resolver directives`,
    );
  }

  const variables: Record<string, SchemaPlanResolverVariable> = {};
  for (const variableDir of directives.filter(
    (d) =>
      d.name.value === 'variable' &&
      mustGetStringArgumentValue(type, field, d, 'subgraph') === source.id,
  )) {
    const variable: SchemaPlanResolverVariable = {
      name: mustGetStringArgumentValue(type, field, variableDir, 'name'),
      select: mustGetStringArgumentValue(type, field, variableDir, 'select'),
    };
    if (variables[variable.name]) {
      throw new Error(
        `Variable "${variable.name}" on type "${type.name}" already defined`,
      );
    }
    variables[variable.name] = variable;
  }

  return {
    url: '', // TODO: source url
    kind: 'fetch',
    operation: mustGetStringArgumentValue(
      type,
      field,
      resolverDir,
      'operation',
    ),
    variables,
  };
}

function mustGetStringArgumentValue(
  type: GraphQLObjectType,
  field: GraphQLField<unknown, unknown, unknown> | null,
  directive: ConstDirectiveNode,
  name: string,
): string {
  const arg = directive.arguments?.find((s) => s.name.value === name)?.value;
  if (!arg) {
    throw new Error(
      `directive @${directive.name.value} on type "${type.name}"${field ? ` at field "${field.name}"` : ''} doesnt have an "${name}" argument`,
    );
  }
  if (arg.kind !== Kind.STRING) {
    throw new Error(
      `directive @${directive.name.value} on type "${type.name}"${field ? ` at field "${field.name}"` : ''} argument "${name}" is not a string`,
    );
  }
  return arg.value;
}
