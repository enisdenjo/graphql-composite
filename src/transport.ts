import { ExecutionResult } from 'graphql';

export interface Transport {
  get(args: TransportGetArgs): Promise<ExecutionResult>;
}

export interface TransportGetArgs {
  query: string;
  variables: Record<string, unknown>;
}

export class TransportHTTP implements Transport {
  constructor(private fetch: typeof globalThis.fetch) {}
  async get({ query, variables }: TransportGetArgs): Promise<ExecutionResult> {
    const res = await this.fetch('http://localhost/graphql', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
        accept: 'application/graphql-response+json, application/json',
      },
      body: JSON.stringify({
        query,
        variables,
      }),
    });
    if (!res.ok) {
      const err = new Error(
        `${res.status} ${res.statusText}\n${await res.text()}\nQuery:${query}\nVariables:${JSON.stringify(variables)}`,
      );
      err.name = 'ResponseError';
      throw err;
    }
    return await res.json();
  }
}
