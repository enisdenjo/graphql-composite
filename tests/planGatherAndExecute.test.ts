import { buildSchema } from 'graphql';
import { describe, expect, it } from 'vitest';
import { execute } from '../src/execute.js';
import { buildResolverOperation, planGather } from '../src/gather.js';
import { planSchema } from '../src/schemaPlan.js';
import { TransportHTTP } from '../src/transport.js';
import { getFixtures } from './utils.js';

describe.each(await getFixtures())(
  'fixture $name',
  ({ fusiongraph, sources, queries }) => {
    const schema = buildSchema(fusiongraph, {
      assumeValid: true,
    });

    it('should plan schema', () => {
      expect(planSchema(schema)).toMatchSnapshot();
    });

    describe.each(queries)('query $name', ({ document, variables }) => {
      it('should plan gather', () => {
        expect(planGather(schema, document)).toMatchSnapshot();
      });

      it('should execute', async () => {
        await expect(
          execute(
            Object.entries(sources).reduce(
              (agg, [name, source]) => ({
                ...agg,
                [name]: new TransportHTTP(source.fetch),
              }),
              {},
            ),
            planGather(schema, document),
            variables,
          ),
        ).resolves.toMatchSnapshot();
      });
    });
  },
);

it.each([
  {
    name: 'storefront',
    operation:
      'query storefront($id: ID!) { storefront(id: $id) { ...__export } }',
    type: 'Storefront',
    fields: ['id', 'name', 'products.upc'],
  },
  {
    name: 'ProductByUpc',
    operation:
      'query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...__export } }',
    type: 'Product',
    fields: [
      'name',
      'manufacturer.products.upc',
      'manufacturer.products.name',
      'manufacturer.id',
    ],
  },
  {
    name: 'ManufacturerById',
    operation:
      'query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...__export } }',
    type: 'Manufacturer',
    fields: ['id', 'name'],
  },
  {
    name: 'ManufacturerNested',
    operation:
      'query ManufacturerNested { manufacturer { nested { deep { ...__export } } } }',
    type: 'Manufacturer',
    fields: ['id', 'name', 'products.manufacturer.location', 'products.name'],
  },
])(
  'should build proper operation and find __export path for $name resolver',
  ({ operation, type, fields }) => {
    expect(buildResolverOperation(operation, type, fields)).toMatchSnapshot();
  },
);
