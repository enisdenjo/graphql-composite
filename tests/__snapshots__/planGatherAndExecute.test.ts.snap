// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixture 'store' > query 'NotBasic' > should execute 1`] = `
{
  "data": {
    "storefront": {
      "id": "cyberport",
      "name": "CyberPort",
      "products": [
        {
          "manufacturer": {
            "id": "manufacturer-of-laptop",
            "name": "TODO: name of manufacturer manufacturer-of-laptop",
            "products": [
              {
                "name": "TODO: name of product laptop",
                "upc": "laptop",
              },
              {
                "name": "TODO: name of product otherthing",
                "upc": "otherthing",
              },
            ],
          },
          "name": "TODO: name of product laptop",
          "upc": "laptop",
        },
        {
          "manufacturer": {
            "id": "manufacturer-of-computer",
            "name": "TODO: name of manufacturer manufacturer-of-computer",
            "products": [
              {
                "name": "TODO: name of product computer",
                "upc": "computer",
              },
              {
                "name": "TODO: name of product otherthing",
                "upc": "otherthing",
              },
            ],
          },
          "name": "TODO: name of product computer",
          "upc": "computer",
        },
        {
          "manufacturer": {
            "id": "manufacturer-of-headset",
            "name": "TODO: name of manufacturer manufacturer-of-headset",
            "products": [
              {
                "name": "TODO: name of product headset",
                "upc": "headset",
              },
              {
                "name": "TODO: name of product otherthing",
                "upc": "otherthing",
              },
            ],
          },
          "name": "TODO: name of product headset",
          "upc": "headset",
        },
      ],
    },
  },
  "explain": [
    {
      "data": {
        "storefront": {
          "id": "cyberport",
          "name": "CyberPort",
          "products": [
            {
              "upc": "laptop",
            },
            {
              "upc": "computer",
            },
            {
              "upc": "headset",
            },
          ],
        },
      },
      "exports": [
        "id",
        "name",
        "products.upc",
      ],
      "includes": [
        {
          "data": {
            "product": {
              "manufacturer": {
                "id": "manufacturer-of-laptop",
                "products": [
                  {
                    "name": "TODO: name of product laptop",
                    "upc": "laptop",
                  },
                  {
                    "name": "TODO: name of product otherthing",
                    "upc": "otherthing",
                  },
                ],
              },
              "name": "TODO: name of product laptop",
            },
          },
          "exports": [
            "name",
            "manufacturer.products.upc",
            "manufacturer.products.name",
            "manufacturer.id",
          ],
          "includes": [
            {
              "data": {
                "manufacturer": {
                  "name": "TODO: name of manufacturer manufacturer-of-laptop",
                },
              },
              "exports": [
                "name",
              ],
              "includes": [],
              "path": [
                "storefront",
                "products",
                0,
                "manufacturer",
              ],
              "query": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } } fragment export on Manufacturer { name }",
              "variables": {
                "Manufacturer_id": "manufacturer-of-laptop",
              },
            },
          ],
          "path": [
            "storefront",
            "products",
            0,
          ],
          "query": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } } fragment export on Product { name manufacturer { products { upc name } id } }",
          "variables": {
            "Product_upc": "laptop",
          },
        },
        {
          "data": {
            "product": {
              "manufacturer": {
                "id": "manufacturer-of-computer",
                "products": [
                  {
                    "name": "TODO: name of product computer",
                    "upc": "computer",
                  },
                  {
                    "name": "TODO: name of product otherthing",
                    "upc": "otherthing",
                  },
                ],
              },
              "name": "TODO: name of product computer",
            },
          },
          "exports": [
            "name",
            "manufacturer.products.upc",
            "manufacturer.products.name",
            "manufacturer.id",
          ],
          "includes": [
            {
              "data": {
                "manufacturer": {
                  "name": "TODO: name of manufacturer manufacturer-of-computer",
                },
              },
              "exports": [
                "name",
              ],
              "includes": [],
              "path": [
                "storefront",
                "products",
                1,
                "manufacturer",
              ],
              "query": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } } fragment export on Manufacturer { name }",
              "variables": {
                "Manufacturer_id": "manufacturer-of-computer",
              },
            },
          ],
          "path": [
            "storefront",
            "products",
            1,
          ],
          "query": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } } fragment export on Product { name manufacturer { products { upc name } id } }",
          "variables": {
            "Product_upc": "computer",
          },
        },
        {
          "data": {
            "product": {
              "manufacturer": {
                "id": "manufacturer-of-headset",
                "products": [
                  {
                    "name": "TODO: name of product headset",
                    "upc": "headset",
                  },
                  {
                    "name": "TODO: name of product otherthing",
                    "upc": "otherthing",
                  },
                ],
              },
              "name": "TODO: name of product headset",
            },
          },
          "exports": [
            "name",
            "manufacturer.products.upc",
            "manufacturer.products.name",
            "manufacturer.id",
          ],
          "includes": [
            {
              "data": {
                "manufacturer": {
                  "name": "TODO: name of manufacturer manufacturer-of-headset",
                },
              },
              "exports": [
                "name",
              ],
              "includes": [],
              "path": [
                "storefront",
                "products",
                2,
                "manufacturer",
              ],
              "query": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } } fragment export on Manufacturer { name }",
              "variables": {
                "Manufacturer_id": "manufacturer-of-headset",
              },
            },
          ],
          "path": [
            "storefront",
            "products",
            2,
          ],
          "query": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } } fragment export on Product { name manufacturer { products { upc name } id } }",
          "variables": {
            "Product_upc": "headset",
          },
        },
      ],
      "path": [
        "storefront",
      ],
      "query": "query storefront($id: ID!) { storefront(id: $id) { ...export } } fragment export on Storefront { id name products { upc } }",
      "variables": {
        "id": "2",
      },
    },
  ],
}
`;

exports[`fixture 'store' > query 'NotBasic' > should plan gather 1`] = `
{
  "operations": [
    {
      "fields": [
        {
          "fields": [
            {
              "isNonNull": true,
              "kind": "scalar",
              "name": "id",
              "type": "ID",
            },
            {
              "isNonNull": true,
              "kind": "scalar",
              "name": "name",
              "type": "String",
            },
            {
              "fields": [
                {
                  "isNonNull": true,
                  "kind": "scalar",
                  "name": "upc",
                  "type": "ID",
                },
                {
                  "isNonNull": true,
                  "kind": "scalar",
                  "name": "name",
                  "type": "String",
                },
                {
                  "fields": [
                    {
                      "fields": [
                        {
                          "isNonNull": true,
                          "kind": "scalar",
                          "name": "upc",
                          "type": "ID",
                        },
                        {
                          "isNonNull": true,
                          "kind": "scalar",
                          "name": "name",
                          "type": "String",
                        },
                      ],
                      "isList": true,
                      "isListItemNonNull": false,
                      "isNonNull": true,
                      "kind": "composite",
                      "name": "products",
                      "type": "Product",
                    },
                    {
                      "isNonNull": true,
                      "kind": "scalar",
                      "name": "name",
                      "type": "String",
                    },
                  ],
                  "isList": false,
                  "isListItemNonNull": false,
                  "isNonNull": false,
                  "kind": "composite",
                  "name": "manufacturer",
                  "type": "Manufacturer",
                },
              ],
              "isList": true,
              "isListItemNonNull": false,
              "isNonNull": true,
              "kind": "composite",
              "name": "products",
              "type": "Product",
            },
          ],
          "isList": false,
          "isListItemNonNull": false,
          "isNonNull": false,
          "kind": "composite",
          "name": "storefront",
          "type": "Storefront",
        },
      ],
      "name": "NotBasic",
      "resolvers": [
        {
          "exports": [
            "id",
            "name",
            "products.upc",
          ],
          "includes": {
            "products": {
              "exports": [
                "name",
                "manufacturer.products.upc",
                "manufacturer.products.name",
                "manufacturer.id",
              ],
              "includes": {
                "manufacturer": {
                  "exports": [
                    "name",
                  ],
                  "includes": {},
                  "kind": "fetch",
                  "operation": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } }",
                  "source": "manufacturers",
                  "type": "Manufacturer",
                  "typeOrField": "Manufacturer",
                  "url": "http://localhost/graphql",
                  "variables": {
                    "Manufacturer_id": {
                      "name": "Manufacturer_id",
                      "select": "id",
                    },
                  },
                },
              },
              "kind": "fetch",
              "operation": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } }",
              "source": "products",
              "type": "Product",
              "typeOrField": "Product",
              "url": "http://localhost/graphql",
              "variables": {
                "Product_upc": {
                  "name": "Product_upc",
                  "select": "upc",
                },
              },
            },
          },
          "kind": "fetch",
          "operation": "query storefront($id: ID!) { storefront(id: $id) { ...export } }",
          "source": "storefronts",
          "type": "Storefront",
          "typeOrField": "storefront",
          "url": "http://localhost/graphql",
          "variables": {},
        },
      ],
      "type": "query",
    },
  ],
  "paths": [
    "storefront",
    "storefront.id",
    "storefront.name",
    "storefront.products",
    "storefront.products.upc",
    "storefront.products.name",
    "storefront.products.manufacturer",
    "storefront.products.manufacturer.products",
    "storefront.products.manufacturer.products.upc",
    "storefront.products.manufacturer.products.name",
    "storefront.products.manufacturer.name",
  ],
  "query": "query NotBasic {
  storefront(id: "2") {
    id
    name
    products {
      upc
      name
      manufacturer {
        products {
          upc
          name
        }
        name
      }
    }
  }
}",
}
`;

exports[`fixture 'store' > should plan schema 1`] = `
{
  "compositeTypes": {
    "Manufacturer": {
      "fields": {
        "id": {
          "name": "id",
          "sources": {
            "manufacturers": {
              "source": "manufacturers",
              "typeOrField": "id",
            },
            "products": {
              "source": "products",
              "typeOrField": "id",
            },
          },
        },
        "name": {
          "name": "name",
          "sources": {
            "manufacturers": {
              "source": "manufacturers",
              "typeOrField": "name",
            },
          },
        },
        "products": {
          "name": "products",
          "sources": {
            "products": {
              "source": "products",
              "typeOrField": "products",
            },
          },
        },
      },
      "name": "Manufacturer",
      "resolvers": {
        "manufacturers": {
          "kind": "fetch",
          "operation": "query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } }",
          "source": "manufacturers",
          "type": "Manufacturer",
          "typeOrField": "Manufacturer",
          "url": "http://localhost/graphql",
          "variables": {
            "Manufacturer_id": {
              "name": "Manufacturer_id",
              "select": "id",
            },
          },
        },
        "products": {
          "kind": "fetch",
          "operation": "query ManufacturerById($Manufacturer_id: ID!) { _manufacturer(id: $Manufacturer_id) { ...export } }",
          "source": "products",
          "type": "Manufacturer",
          "typeOrField": "Manufacturer",
          "url": "http://localhost/graphql",
          "variables": {
            "Manufacturer_id": {
              "name": "Manufacturer_id",
              "select": "id",
            },
          },
        },
      },
    },
    "Product": {
      "fields": {
        "manufacturer": {
          "name": "manufacturer",
          "sources": {
            "products": {
              "source": "products",
              "typeOrField": "manufacturer",
            },
          },
        },
        "name": {
          "name": "name",
          "sources": {
            "products": {
              "source": "products",
              "typeOrField": "name",
            },
          },
        },
        "price": {
          "name": "price",
          "sources": {
            "products": {
              "source": "products",
              "typeOrField": "price",
            },
          },
        },
        "upc": {
          "name": "upc",
          "sources": {
            "products": {
              "source": "products",
              "typeOrField": "upc",
            },
            "storefronts": {
              "source": "storefronts",
              "typeOrField": "upc",
            },
          },
        },
      },
      "name": "Product",
      "resolvers": {
        "products": {
          "kind": "fetch",
          "operation": "query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } }",
          "source": "products",
          "type": "Product",
          "typeOrField": "Product",
          "url": "http://localhost/graphql",
          "variables": {
            "Product_upc": {
              "name": "Product_upc",
              "select": "upc",
            },
          },
        },
        "storefronts": {
          "kind": "fetch",
          "operation": "TODO",
          "source": "storefronts",
          "type": "Product",
          "typeOrField": "Product",
          "url": "http://localhost/graphql",
          "variables": {
            "Product_upc": {
              "name": "Product_upc",
              "select": "upc",
            },
          },
        },
      },
    },
    "Storefront": {
      "fields": {
        "id": {
          "name": "id",
          "sources": {
            "storefronts": {
              "source": "storefronts",
              "typeOrField": "id",
            },
          },
        },
        "name": {
          "name": "name",
          "sources": {
            "storefronts": {
              "source": "storefronts",
              "typeOrField": "name",
            },
          },
        },
        "products": {
          "name": "products",
          "sources": {
            "storefronts": {
              "source": "storefronts",
              "typeOrField": "products",
            },
          },
        },
      },
      "name": "Storefront",
      "resolvers": {
        "storefronts": {
          "kind": "fetch",
          "operation": "query StorefrontById($Storefront_id: ID!) { storefront(id: $Storefront_id) { ...export } }",
          "source": "storefronts",
          "type": "Storefront",
          "typeOrField": "Storefront",
          "url": "http://localhost/graphql",
          "variables": {
            "Storefront_id": {
              "name": "Storefront_id",
              "select": "id",
            },
          },
        },
      },
    },
  },
  "operations": {
    "query": {
      "fields": {
        "_manufacturer": {
          "name": "_manufacturer",
          "resolvers": {
            "products": {
              "kind": "fetch",
              "operation": "query _manufacturer($id: ID!) { _manufacturer(id: $id) { ...export } }",
              "source": "products",
              "type": "Manufacturer",
              "typeOrField": "_manufacturer",
              "url": "http://localhost/graphql",
              "variables": {},
            },
          },
        },
        "manufacturer": {
          "name": "manufacturer",
          "resolvers": {
            "manufacturers": {
              "kind": "fetch",
              "operation": "query manufacturer($id: ID!) { manufacturer(id: $id) { ...export } }",
              "source": "manufacturers",
              "type": "Manufacturer",
              "typeOrField": "manufacturer",
              "url": "http://localhost/graphql",
              "variables": {},
            },
          },
        },
        "product": {
          "name": "product",
          "resolvers": {
            "products": {
              "kind": "fetch",
              "operation": "query product($upc: ID!) { product(upc: $upc) }",
              "source": "products",
              "type": "Product",
              "typeOrField": "product",
              "url": "http://localhost/graphql",
              "variables": {},
            },
          },
        },
        "storefront": {
          "name": "storefront",
          "resolvers": {
            "storefronts": {
              "kind": "fetch",
              "operation": "query storefront($id: ID!) { storefront(id: $id) { ...export } }",
              "source": "storefronts",
              "type": "Storefront",
              "typeOrField": "storefront",
              "url": "http://localhost/graphql",
              "variables": {},
            },
          },
        },
      },
      "name": "query",
    },
  },
}
`;

exports[`should build proper query for 'manufacturers' resolver 1`] = `"query ManufacturerById($Manufacturer_id: ID!) { manufacturer(id: $Manufacturer_id) { ...export } } fragment export on Manufacturer { name }"`;

exports[`should build proper query for 'products' resolver 1`] = `"query ProductByUpc($Product_upc: ID!) { product(upc: $Product_upc) { ...export } } fragment export on Product { name manufacturer { products { upc name } id } }"`;

exports[`should build proper query for 'storefronts' resolver 1`] = `"query storefront($id: ID!) { storefront(id: $id) { ...export } } fragment export on Storefront { id name products { upc } }"`;
